velox/type/TimestampConversion.cpp

```diff
@@ -593,6 +613,26 @@ Status daysSinceEpochFromWeekDate(
   return Status::OK();
 }
 
+Status daysSinceEpochFromWeekOfMonthDate(
+    int32_t year,
+    int32_t month,
+    int32_t weekOfMonth,
+    int32_t dayOfWeek,
+    int64_t& out) {
+  if (!isValidWeekOfMonthDate(year, month, weekOfMonth, dayOfWeek)) {
+    return Status::UserError(
+        "Date out of range: {}-{}-{}-{}", year, month, weekOfMonth, dayOfWeek);
+  }
+  int64_t daysSinceEpochOfFirstDayOfMonth;
+  VELOX_RETURN_NOT_OK(
+      daysSinceEpochFromDate(year, month, 1, daysSinceEpochOfFirstDayOfMonth));
+  int32_t firstDayOfWeek =
```

Comment:
In Java's SimpleDateFormat, "invalid" weekOfMonth and dayOfWeek etc. are allowed. If the date falls outside the valid range, the week or day from the preceding or following month's will be used. For example, if weekOfMonth is 5 but the current month only has 4 weeks (such as February), the first week of March will be considered as the 5th week of February. Similarly, if weekOfMonth is 5 and dayOfWeek is 7, the corresponding day from the next month will be used.

Reply:
Thank you for clarifying. Let's update method comments to explain this as it is not obvious. Do we already have tests for these cases? If not, let's make sure to add.

Reply:
I added this comments for clarifying and two tests for cases that days in previous/next month.

> We treat days of the previous or next months as a part of the specified WEEK_OF_MONTH. For example, if
 weekOfMonth is 5 but the current month only has 4 weeks (such as February),the first week of March will be considered as the 5th week of February.

Reply:
I'm seeing that DATE type is based on INTEGER type, hence, allows only int32_t number of days since epoch. If that's the case, then, perhaps, daysSinceEpochFromDate should return int32_t, not int64_t.

Reply:
This PR is intended for date parsing, and the DateTimeFormatter::parse method returns a Timestamp. A Timestamp is based on int64_t seconds and can represent up to 106,751,991,167 days, which is calculated as (2^63 - 1) / 1000 / 86400. This value is greater than the maximum value of an int32_t. Therefore, I think that daysSinceEpochFromDate should return an int64_t.